### PR TITLE
[Lucas Johnson] iP

### PR DESCRIPTION
**Pull request by eman-kom**
# This is a Duke Bot

>Do today, not ~tomorrow~ 👈 😤

- Create **Todo** Tasks
- Create **Event** Tasks
- Create **Deadline** Tasks

1. You can *find* tasks
2. You can *save* tasks
3. You can *delete* tasks

- [x] Mark tasks as completed
- [ ] Mark tasks as not completed

```java
System.out.println("System.out.println.");
```

Use `git status` to get the status of your git.

[Link to the docs page](https://eman-kom.github.io/ip/).

### CODE REVIEW

File: src/main/java/seedu/storage/Storage.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Scanner;
+
+public class Storage {
+
+    private final File file;
+    private final SimpleDateFormat ft = new SimpleDateFormat("dd MMM yyy h.mma");
```
Comment (Wei-Xinn) [inline src/main/java/seedu/storage/Storage.java]:
```suggestion
    private final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("dd MMM yyy h.mma");
```
Consider changing naming variable to match naming convention for constants.

File: src/main/java/seedu/duke/Parser.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        int idx;
+        Task newTask;
+        String[] parsedTask;
+        switch (cmd.toLowerCase().trim()) {
+
+        case "list":
+            System.out.println(tasks.toString());
+            break;
+
+        case "mark":
```
Comment (Wei-Xinn) [inline src/main/java/seedu/duke/Parser.java]:
```suggestion
        case "list":
            System.out.println(tasks.toString());
            break;
        case "mark":
```
I am not very sure about this, but the coding standard should be that there shouldn't be a line break after "break;"

File: src/main/java/seedu/duke/Duke.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                if (parser.checkQuit(cmd[0].trim())) {
+                    isQuit = true;
+                } else {
+                    String task = cmd.length < 2 ? "" : cmd[1];
+                    parser.runCmd(cmd[0].trim(), task.trim(), tasks, ui, storage);
+                }
+
+            } catch (DukeException | IOException e) {
+                ui.showError(e.toString());
+            }
```
Comment (Wei-Xinn) [inline src/main/java/seedu/duke/Duke.java]:
```suggestion
            try {
                String input = ui.getCommand();
                String[] cmd = parser.splitCmd(input);

                if (parser.checkQuit(cmd[0].trim())) {
                    isQuit = true;
                } else {
                    String task = cmd.length < 2 ? "" : cmd[1];
                    parser.runCmd(cmd[0].trim(), task.trim(), tasks, ui, storage);
                }
            } catch (DukeException | IOException e) {
                ui.showError(e.toString());
            }
```
Good line breaks to make it the logic clear and readable!
But I do think that there is a need for the line breaks after try and before catch

File: src/main/java/seedu/duke/Parser.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Date;
+
+public class Parser {
+
+    private final String[] cmdTypes = {"find", "list", "delete", "unmark", "mark", "todo", "event", "deadline", "bye"};
```
Comment (Wei-Xinn) [inline src/main/java/seedu/duke/Parser.java]:
```suggestion
    private final String[] CMD_TYPES = {"find", "list", "delete", "unmark", "mark", "todo", "event", "deadline", "bye"};
```
Consider changing variable name to match constant naming convention.

File: src/main/java/seedu/duke/Parser.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            throw new DukeException("\"" + keyWord + "\" Not Found");
+        } else {
+            if (splitTask[0].equals("") && splitTask[1].equals("")) {
+                throw new DukeException("No task in instruction");
+            } else if (splitTask[0].equals("")) {
+                throw new DukeException("Description of task not found");
+            } else if (splitTask[1].equals("")) {
+                throw new DukeException("Datetime of task not found");
+            }
+        }
```
Comment (Wei-Xinn) [inline src/main/java/seedu/duke/Parser.java]: Concise and readable!

File: src/main/java/seedu/task/Event.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package task;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class Event extends Task {
+
+    private final Date duration;
+    private final SimpleDateFormat ft = new SimpleDateFormat("dd MMM yyy h.mma");
```
Comment (Wei-Xinn) [inline src/main/java/seedu/task/Event.java]: Same as above. Maybe consider changing variable name to match constant naming convention

File: src/main/java/seedu/task/Task.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
+package task;
+
+public class Task {
+
+    private final String description;
+    private boolean completed;
```
Comment (Wei-Xinn) [inline src/main/java/seedu/task/Task.java]:
```suggestion
    private boolean isCompleted;
```
Would isCompleted be a more intuitive vairable name?

File: src/main/java/seedu/duke/Ui.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
+package seedu.duke;
+
+import task.Task;
+
+import java.util.Scanner;
+
+public class Ui {
+
+    private final Scanner sc;
```
Comment (hanqinilnix) [inline src/main/java/seedu/duke/Ui.java]: Constant variables should be capitalised according to the coding standards. I noticed similar issues in other areas as well.

File: src/main/java/seedu/duke/Duke.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+                String input = ui.getCommand();
+                String[] cmd = parser.splitCmd(input);
+
+                if (parser.checkQuit(cmd[0].trim())) {
+                    isQuit = true;
+                } else {
+                    String task = cmd.length < 2 ? "" : cmd[1];
+                    parser.runCmd(cmd[0].trim(), task.trim(), tasks, ui, storage);
+                }
```
Comment (hanqinilnix) [inline src/main/java/seedu/duke/Duke.java]: Wouldn't it be better if you encapsulated this in the Parser class?

File: src/main/java/seedu/storage/Storage.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+     * @throws IOException file error
+     * @throws ParseException date not correct format
+     */
+    public ArrayList<Task> read() throws IOException, ParseException {
+
+        Scanner sc = new Scanner(this.file);
+        ArrayList<Task> taskList = new ArrayList<>();
+
+        while (sc.hasNext()) {
+            String[] task = sc.nextLine().trim().split("\t");
```
Comment (hanqinilnix) [inline src/main/java/seedu/storage/Storage.java]:
```suggestion
            String[] taskSegments = sc.nextLine().trim().split("\t");
```
Perhaps you can nouns that reflect the variable better. Plural form is also be used when representing a collection of objects.

File: src/main/java/seedu/task/Deadline.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        this.deadline = deadline;
+    }
+
+    public Deadline(String description, Date deadline, boolean completed) {
+        super(description, completed);
+        this.deadline = deadline;
+    }
+
+    /**
+     * @inheritDoc
```
Comment (hanqinilnix) [inline src/main/java/seedu/task/Deadline.java]: Interesting annotation

File: src/main/java/seedu/task/Task.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    public Task(String description) {
+        this(description, false);
+    }
+
+    /**
+     * Creates new Task object
+     * @param description of task
+     * @param completed task is completed or not
+     */
+    public Task(String description, boolean completed) {
```
Comment (hanqinilnix) [inline src/main/java/seedu/task/Task.java]: Great use overloading, I didn't think of this

**Review by hanqinilnix — COMMENTED** [review]
Let's do this

**Pull request closed without merging**